// @rc-component/util/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "KeyCode", {
  enumerable: true,
  get: function () {
    return _KeyCode.default;
  }
});
Object.defineProperty(exports, "canUseDom", {
  enumerable: true,
  get: function () {
    return _canUseDom.default;
  }
});
Object.defineProperty(exports, "composeRef", {
  enumerable: true,
  get: function () {
    return _ref.composeRef;
  }
});
Object.defineProperty(exports, "contains", {
  enumerable: true,
  get: function () {
    return _contains.default;
  }
});
Object.defineProperty(exports, "fillRef", {
  enumerable: true,
  get: function () {
    return _ref.fillRef;
  }
});
Object.defineProperty(exports, "get", {
  enumerable: true,
  get: function () {
    return _get.default;
  }
});
Object.defineProperty(exports, "getDOM", {
  enumerable: true,
  get: function () {
    return _findDOMNode.getDOM;
  }
});
Object.defineProperty(exports, "getFocusNodeList", {
  enumerable: true,
  get: function () {
    return _focus.getFocusNodeList;
  }
});
Object.defineProperty(exports, "getId", {
  enumerable: true,
  get: function () {
    return _useId.getId;
  }
});
Object.defineProperty(exports, "getNodeRef", {
  enumerable: true,
  get: function () {
    return _ref.getNodeRef;
  }
});
Object.defineProperty(exports, "getScrollBarSize", {
  enumerable: true,
  get: function () {
    return _getScrollBarSize.default;
  }
});
Object.defineProperty(exports, "getShadowRoot", {
  enumerable: true,
  get: function () {
    return _shadow.getShadowRoot;
  }
});
Object.defineProperty(exports, "getTargetScrollBarSize", {
  enumerable: true,
  get: function () {
    return _getScrollBarSize.getTargetScrollBarSize;
  }
});
Object.defineProperty(exports, "injectCSS", {
  enumerable: true,
  get: function () {
    return _dynamicCSS.injectCSS;
  }
});
Object.defineProperty(exports, "isDOM", {
  enumerable: true,
  get: function () {
    return _findDOMNode.isDOM;
  }
});
Object.defineProperty(exports, "isEqual", {
  enumerable: true,
  get: function () {
    return _isEqual.default;
  }
});
Object.defineProperty(exports, "isMobile", {
  enumerable: true,
  get: function () {
    return _isMobile.default;
  }
});
Object.defineProperty(exports, "isVisible", {
  enumerable: true,
  get: function () {
    return _isVisible.default;
  }
});
Object.defineProperty(exports, "lockFocus", {
  enumerable: true,
  get: function () {
    return _focus.lockFocus;
  }
});
Object.defineProperty(exports, "merge", {
  enumerable: true,
  get: function () {
    return _set.merge;
  }
});
Object.defineProperty(exports, "mergeProps", {
  enumerable: true,
  get: function () {
    return _mergeProps.default;
  }
});
Object.defineProperty(exports, "mergeWith", {
  enumerable: true,
  get: function () {
    return _set.mergeWith;
  }
});
Object.defineProperty(exports, "noteOnce", {
  enumerable: true,
  get: function () {
    return _warning.noteOnce;
  }
});
Object.defineProperty(exports, "omit", {
  enumerable: true,
  get: function () {
    return _omit.default;
  }
});
Object.defineProperty(exports, "pickAttrs", {
  enumerable: true,
  get: function () {
    return _pickAttrs.default;
  }
});
Object.defineProperty(exports, "proxyObject", {
  enumerable: true,
  get: function () {
    return _proxyObject.default;
  }
});
Object.defineProperty(exports, "raf", {
  enumerable: true,
  get: function () {
    return _raf.default;
  }
});
Object.defineProperty(exports, "removeCSS", {
  enumerable: true,
  get: function () {
    return _dynamicCSS.removeCSS;
  }
});
Object.defineProperty(exports, "set", {
  enumerable: true,
  get: function () {
    return _set.default;
  }
});
Object.defineProperty(exports, "supportNodeRef", {
  enumerable: true,
  get: function () {
    return _ref.supportNodeRef;
  }
});
Object.defineProperty(exports, "supportRef", {
  enumerable: true,
  get: function () {
    return _ref.supportRef;
  }
});
Object.defineProperty(exports, "toArray", {
  enumerable: true,
  get: function () {
    return _toArray.default;
  }
});
Object.defineProperty(exports, "triggerFocus", {
  enumerable: true,
  get: function () {
    return _focus.triggerFocus;
  }
});
Object.defineProperty(exports, "updateCSS", {
  enumerable: true,
  get: function () {
    return _dynamicCSS.updateCSS;
  }
});
Object.defineProperty(exports, "useComposeRef", {
  enumerable: true,
  get: function () {
    return _ref.useComposeRef;
  }
});
Object.defineProperty(exports, "useControlledState", {
  enumerable: true,
  get: function () {
    return _useControlledState.default;
  }
});
Object.defineProperty(exports, "useEvent", {
  enumerable: true,
  get: function () {
    return _useEvent.default;
  }
});
Object.defineProperty(exports, "useId", {
  enumerable: true,
  get: function () {
    return _useId.default;
  }
});
Object.defineProperty(exports, "useLayoutEffect", {
  enumerable: true,
  get: function () {
    return _useLayoutEffect.default;
  }
});
Object.defineProperty(exports, "useLayoutUpdateEffect", {
  enumerable: true,
  get: function () {
    return _useLayoutEffect.useLayoutUpdateEffect;
  }
});
Object.defineProperty(exports, "useLockFocus", {
  enumerable: true,
  get: function () {
    return _focus.useLockFocus;
  }
});
Object.defineProperty(exports, "useMemo", {
  enumerable: true,
  get: function () {
    return _useMemo.default;
  }
});
Object.defineProperty(exports, "useMergedState", {
  enumerable: true,
  get: function () {
    return _useMergedState.default;
  }
});
Object.defineProperty(exports, "useState", {
  enumerable: true,
  get: function () {
    return _useState.default;
  }
});
Object.defineProperty(exports, "useSyncState", {
  enumerable: true,
  get: function () {
    return _useSyncState.default;
  }
});
Object.defineProperty(exports, "warning", {
  enumerable: true,
  get: function () {
    return _warning.default;
  }
});
var _useEvent = _interopRequireDefault(require("./hooks/useEvent"));
var _useMergedState = _interopRequireDefault(require("./hooks/useMergedState"));
var _useControlledState = _interopRequireDefault(require("./hooks/useControlledState"));
var _useId = _interopRequireWildcard(require("./hooks/useId"));
var _useLayoutEffect = _interopRequireWildcard(require("./hooks/useLayoutEffect"));
var _useMemo = _interopRequireDefault(require("./hooks/useMemo"));
var _useState = _interopRequireDefault(require("./hooks/useState"));
var _useSyncState = _interopRequireDefault(require("./hooks/useSyncState"));
var _ref = require("./ref");
var _canUseDom = _interopRequireDefault(require("./Dom/canUseDom"));
var _contains = _interopRequireDefault(require("./Dom/contains"));
var _dynamicCSS = require("./Dom/dynamicCSS");
var _findDOMNode = require("./Dom/findDOMNode");
var _focus = require("./Dom/focus");
var _isVisible = _interopRequireDefault(require("./Dom/isVisible"));
var _shadow = require("./Dom/shadow");
var _KeyCode = _interopRequireDefault(require("./KeyCode"));
var _getScrollBarSize = _interopRequireWildcard(require("./getScrollBarSize"));
var _isEqual = _interopRequireDefault(require("./isEqual"));
var _isMobile = _interopRequireDefault(require("./isMobile"));
var _omit = _interopRequireDefault(require("./omit"));
var _pickAttrs = _interopRequireDefault(require("./pickAttrs"));
var _proxyObject = _interopRequireDefault(require("./proxyObject"));
var _raf = _interopRequireDefault(require("./raf"));
var _toArray = _interopRequireDefault(require("./Children/toArray"));
var _mergeProps = _interopRequireDefault(require("./mergeProps"));
var _get = _interopRequireDefault(require("./utils/get"));
var _set = _interopRequireWildcard(require("./utils/set"));
var _warning = _interopRequireWildcard(require("./warning"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
